rtt/extras/FileDescriptorActivity.cpp

```diff
@@ -285,7 +290,15 @@ bool FileDescriptorActivity::trigger()
 
 bool FileDescriptorActivity::timeout()
 {
-    return false;
+    if (isActive() ) {
+        { RTT::os::MutexLock lock(m_command_mutex);
+            m_user_timeout = true;
+        }
+        int unused; (void)unused;
```

Comment:
Done factoring out the code ... good catch.

As for timeout(), I can't because `Activity` does not have any and to be honest that's not my part of the RTT code - I don't know what `timeout()` is meant to be. I only followed the code and found out what was the problem. Given that Activity does trigger on timeout(), I deduced that it was probably not a bad thing that FDA does as well.

Hopefully if I have got something wrong the RTT people will chime in. I'd suggest to drop a line on the RTT pull request to ask for more documentation on Activity.

On the Rock side, it is definitely an improvement. The current behavior rather badly breaks a common usage of fd_driven.

Reply:
I saw that it is missing on Activity, but there is minimal documentation on the return type see [FileDescriptorActivity::timeout](https://github.com/rock-core/rtt/blob/e33e084c36f11381dbd2edd96fd498c45b741ec5/rtt/extras/FileDescriptorActivity.hpp#L309) L309  and this would be at the very least now misleading.

Reply:
I've done my best ... it's basically what I could find out reading the code.

The RunnableInterface API is very unclear in that respect. `work` has been introduced recently, aiming at replacing `step`, but there's no guidance on whether one should use step or work. What I did find out is that ExecutionEngine#step is a no-op, so if you **are** using step, it's going to do nothing.

Reply:
much better, thx for the effort